TDL15-CloudFormation-Jenkins/cf-deploy-jenkins.yaml

```diff
@@ -0,0 +1,245 @@
+        - !Ref 'AWS::Region'
+        - AMI
+      # Pull the Intance Type, Subnet, KeyName, etc from the Parameters we defined earlier
+      InstanceType: !Ref InstanceType
+      NetworkInterfaces:
+        - AssociatePublicIpAddress: true
+          DeviceIndex: 0
+          DeleteOnTermination: true
+          SubnetId: !Ref SubnetId
+          # This defines the SG that we will place on the ENI, We will create the SG after the instance resource definition
+          GroupSet: 
+            - !Ref ServerSecurityGroup
+      KeyName: !Ref KeyName
+      Tags:
+       - Key: Name
+         Value: !Ref InstanceTagName
+      # Use the user data to install Jenkins and other tools.
+      UserData:
+        Fn::Base64: !Sub |   
+          #!/bin/bash -x
+          # Wait for additional EBS volume attached to JenkinsInstance
+          while true; do lsblk | grep xvdb && break; sleep 10; done
+          # Attach the additional EBS volume and mount the volume in /var/lib/Jenkins
+          sudo mkdir -p /var/lib/jenkins  
+          if [ -z ${SnapshotId} ] ;then echo -e "o\nn\np\n1\n\n\nw" | sudo fdisk /dev/sdb; sleep 3; sudo mkfs.ext4 /dev/sdb1; sudo e2label /dev/sdb1 JENKINS; fi
+          echo -e "LABEL=JENKINS     /var/lib/jenkins    ext4   defaults 0 0" >> /etc/fstab
+          mount -a
+          # Ensure that your software packages are up to date on your instance
+          sudo yum update –y
+          # install Jenkins
+          sudo yum install -y java-1.8.0-openjdk-devel
+          # Enable the Jenkins repository. Download the repo file and import the GPG key
+          sudo yum install -y wget
+          # sudo rpm -ivh https://dl.fedoraproject.org/pub/epel/epel-release-latest-7.noarch.rpm
+          sudo wget -O /etc/yum.repos.d/jenkins.repo http://pkg.jenkins-ci.org/redhat/jenkins.repo
+          sudo yum clean all
+          sudo rpm --import https://pkg.jenkins.io/redhat/jenkins.io.key
+          # make jenkins dirs:
+          sudo mkdir -p /home/jenkins /var/lib/jenkins/.ssh /var/cache/jenkins/war /var/log/jenkins
+          # Install the latest stable version of Jenkins
+          sudo yum install -y jenkins
+          # check whether Jenkins is running, type:
+          # systemctl status jenkins
+          # check jenkins login: 
+          sudo useradd -d /var/lib/jenkins jenkins # login dir
+          usermod -s /bin/bash jenkins # shell use dir
+          # set_jenkins_port  
+          sudo sed -i "s/JENKINS_PORT=\"8080\"/JENKINS_PORT=\"${JenkinsPort}\"/g" /etc/sysconfig/jenkins
+          # sshconfig: 
+          sudo echo -e "Host *\n\tStrictHostKeyChecking no\n" >> /var/lib/jenkins/.ssh/config
+          # ssh-keyscan -H hostname.com >> /var/lib/jenkins/.ssh/known_hosts
+          # Generate SSH key: 
+          if [ -z ${SnapshotId} ] ;then sudo ssh-keygen -t rsa -b 2048 -C jenkins@jenkins -N "" -f /var/lib/jenkins/.ssh/id_rsa; sudo cat /var/lib/jenkins/.ssh/id_rsa.pub > /var/lib/jenkins/.ssh/authorized_keys; fi
```

Comment:
Why do we need to create SSH key for Jenkins user? Jenkins is a service account.
Ops team should ssh to the jenkins box using their own user then sudo to jenkjins user for troubleshooting purpose.

Reply:
For further integrate with GitHub. we can use SSH public key, Sure on GitHub set token is also ok.